File: src/streambench/stormbench/src/main/java/com/intel/hibench/streambench/storm/micro/DistinctCount.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.intel.hibench.streambench.storm.micro;

import com.intel.hibench.streambench.storm.topologies.SingleSpoutTops;
import com.intel.hibench.streambench.storm.util.BenchLogUtil;
import com.intel.hibench.streambench.storm.util.StormBenchConfig;
import org.apache.storm.topology.BasicOutputCollector;
import org.apache.storm.topology.OutputFieldsDeclarer;
import org.apache.storm.topology.TopologyBuilder;
import org.apache.storm.topology.base.BaseBasicBolt;
import org.apache.storm.tuple.Fields;
import org.apache.storm.tuple.Tuple;
import org.apache.storm.tuple.Values;

import java.util.HashSet;
import java.util.Set;

public class DistinctCount extends SingleSpoutTops {

  public DistinctCount(StormBenchConfig config) {
    super(config);
  }

  @Override
  public void setBolts(TopologyBuilder builder) {
    builder.setBolt("sketch", new ProjectStreamBolt(config.fieldIndex, config.separator),
            config.boltThreads).shuffleGrouping("spout");
    builder.setBolt("distinct", new TotalDistinctCountBolt(),
            config.boltThreads).fieldsGrouping("sketch", new Fields("field"));
  }

  private static class TotalDistinctCountBolt extends BaseBasicBolt {
    Set<String> set = new HashSet<String>();

    @Override
    public void execute(Tuple tuple, BasicOutputCollector collector) {
      String word = tuple.getString(0); //FIXME: always pick up index 0? should be configurable according to sparkstream's version
      set.add(word);
      BenchLogUtil.logMsg("Distinct count:" + set.size());
      collector.emit(new Values(set.size()));
    }

    @Override
    public void declareOutputFields(OutputFieldsDeclarer declarer) {
      declarer.declare(new Fields("size"));
    }
  }


}

File: src/streambench/gearpumpbench/src/main/scala/com/intel/hibench/streambench/gearpump/application/DistinctCount.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.intel.hibench.streambench.gearpump.application

import com.intel.hibench.streambench.common.TestCase
import com.intel.hibench.streambench.gearpump.source.SourceProvider
import com.intel.hibench.streambench.gearpump.task.{TotalDistinctCount, Project}
import com.intel.hibench.streambench.gearpump.util.GearpumpConfig
import org.apache.gearpump.cluster.UserConfig
import org.apache.gearpump.partitioner.{HashPartitioner, ShufflePartitioner}
import org.apache.gearpump.streaming.{Processor, StreamApplication}
import org.apache.gearpump.util.Graph
import org.apache.gearpump.util.Graph._

class DistinctCount(conf: GearpumpConfig)(implicit sourceProvider: SourceProvider) extends BasicApplication(conf)  {
  override val benchName = TestCase.DistinctCount

  override def application(benchConfig: UserConfig): StreamApplication = {
    val source = getSource()
    val shufflePartitioner = new ShufflePartitioner
    val hashPartitioner = new HashPartitioner
    val projection = Processor[Project](conf.parallelism)
    val count = Processor[TotalDistinctCount](conf.parallelism)
    val graph = Graph(source ~ shufflePartitioner ~> projection ~ hashPartitioner ~> count)
    StreamApplication("distinctCount", graph, benchConfig)
  }
}
